sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt

```diff
@@ -0,0 +1,331 @@
+            decryptedRecord.customCreationDate,
+            encryptedDataKey,
+            encryptedAttachmentKey,
+            decryptedRecord.modelVersion
+        )
+    }
+
+    private fun validateRecord(encryptedResource: String, version: Int) {
+        if (encryptedResource.isEmpty()) {
+            throw CoreRuntimeException.InternalFailure()
+        }
+
+        if (!modelVersion.isModelVersionSupported(version)) {
+            throw DataValidationException.ModelVersionNotSupported("Please update SDK to latest version!")
+        }
+    }
+
+    private fun getCommonKey(commonKeyId: String, userId: String): GCKey {
+        val commonKeyStored = cryptoService.hasCommonKey(commonKeyId)
+        return if (commonKeyStored) {
+            cryptoService.getCommonKeyById(commonKeyId)
+        } else {
+            // TODO: This should be in a different Service
+            val commonKeyResponse = apiService.fetchCommonKey(
+                alias,
+                userId,
+                commonKeyId
+            ).blockingGet()
+
+            cryptoService.asymDecryptSymetricKey(
+                cryptoService.fetchGCKeyPair().blockingGet(),
+                commonKeyResponse.commonKey
+            ).blockingGet().also {
+                cryptoService.storeCommonKey(commonKeyId, it)
+            }
+        }
+    }
+
+    private fun decryptKey(
+        commonKey: GCKey,
+        key: NetworkModelContract.EncryptedKey
+    ): GCKey {
+        return cryptoService.symDecryptSymmetricKey(
+            commonKey,
+            key
+        ).blockingGet()
+    }
+
+    private fun decryptKeys(
+        commonKey: GCKey,
+        encryptedDataKey: NetworkModelContract.EncryptedKey,
+        encryptedAttachmentKey: NetworkModelContract.EncryptedKey?
+    ): Pair<GCKey, GCKey?> {
+        val dataKey = decryptKey(commonKey, encryptedDataKey)
+
+        return if (encryptedAttachmentKey is NetworkModelContract.EncryptedKey) {
+            Pair(dataKey, decryptKey(commonKey, encryptedAttachmentKey))
+        } else {
+            Pair(dataKey, null)
+        }
+    }
+
+    override fun <T : Any> decrypt(
+        encryptedRecord: NetworkModelContract.EncryptedRecord,
+        userId: String
+    ): NetworkModelContract.DecryptedBaseRecord<T> {
+        validateRecord(encryptedRecord.encryptedBody, encryptedRecord.modelVersion)
+
+        val (tags, annotations) = tagEncryptionService.decryptTagsAndAnnotations(
+            encryptedRecord.encryptedTags
+        )
+        val commonKey = getCommonKey(encryptedRecord.commonKeyId, userId)
+        val (dataKey, attachmentKey) = decryptKeys(
+            commonKey,
+            encryptedRecord.encryptedDataKey,
+            encryptedRecord.encryptedAttachmentsKey
+        )
+
+        return buildRecord(
+            encryptedRecord.identifier,
+            fhirService.decryptResource(dataKey, tags, encryptedRecord.encryptedBody),
```

Comment:
is there any way that you can avoid using the `fhirService` here? seems very out of place to me, If this is a record(resource)cryptor

Reply:
No, we cannot avoid that for now. The FhirService is actually more a ResourceCryptoService. We'll come to this once we bubbled up to the corresponding part in the SDK.

Reply:
I think the root cause, is more like, then we should slice out this whole mapping record stuff out (same goes for the `fromRecord` method - it will live in the future somewhere else). However I did not it here since it causes much less effort, when finally Koin is in place.

Reply:
Totally forgot, we'll get rid of the entire mappings of the resources internally, since they will be wrapped. This will in total improve the whole resource handling as well as the readability, but this something which will be seen later in time, since there’s  still much more in queue before.
However I made a follow up to address the overall naming issues - #128.